components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimespansStorageEngine.kt

```diff
@@ -0,0 +1,236 @@
+            return
+        }
+
+        synchronized(this) {
+            uncommittedStartTimes[timespanName] = getElapsedNanos()
+        }
+    }
+
+    /**
+     * Stop tracking time for the provided metric. Add the elapsed time to the time currently
+     * stored in the metric. This will record an error if no [start] was called.
+     *
+     * @param metricData the metric information for the timespan
+     * @param timeUnit the time unit we want the data in when snapshotting
+     */
+    @Synchronized
+    fun stopAndSum(
+        metricData: CommonMetricData,
+        timeUnit: TimeUnit
+    ) {
+        // TODO report errors if not tracking time through internal metrics. See bug 1499761.
+        // Look for the start time: if it's there, commit the timespan.
+        val timespanName = getStoredName(metricData)
+        uncommittedStartTimes.remove(timespanName)?.let { startTime ->
+            val elapsedNanos = getElapsedNanos() - startTime
+
+            // Store the time unit: we'll need it when snapshotting.
+            timeUnitsMap[timespanName] = timeUnit
+
+            // Use a custom combiner to sum the new timespan to the one already stored. We
+            // can't adjust the time unit before storing so that we still allow for values
+            // lower than the desired time unit to accumulate.
+            super.recordScalar(metricData, elapsedNanos, timeUnit) { currentValue, newValue ->
+                currentValue?.let {
+                    it + newValue
+                } ?: newValue
+            }
+        }
+    }
+
+    /**
+     * Abort a previous [start] call. No error is recorded if no [start] was called.
+     *
+     * @param metricData the metric information for the timespan
+     */
+    @Synchronized
+    fun cancel(metricData: CommonMetricData) {
+        uncommittedStartTimes.remove(getStoredName(metricData))
+    }
+
+    /**
+     * Get a snapshot of the stored timespans and adjust it to the desired time units.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clear the requested store. Not that only
+     *        metrics stored with a lifetime of [Lifetime.Ping] will be cleared.
+     *
+     * @return the [Long] recorded in the requested store
+     */
+    @Synchronized
+    override fun getSnapshot(storeName: String, clearStore: Boolean): GenericDataStorage<Long>? {
+        val adjustedData = super.getSnapshot(storeName, clearStore)?.mapValues {
+            // Convert to the expected time unit.
+            if (it.key !in timeUnitsMap) {
+                logger.error("Can't find the time unit for ${it.key}. Reporting raw value.")
+            }
+
+            timeUnitsMap[it.key]?.let { timeUnit ->
+                getAdjustedTime(timeUnit, it.value)
+            } ?: it.value
+        }
+
+        // Clear the time unit map if needed: we need to check all the stores
+        // for all the lifetimes.
+        if (clearStore) {
```

Comment:
I don't know if we need to address this now...  We don't strictly need to clear these, since there is a pretty small maximum possible size (the number of timespan metrics defined at build time), and the memory churn may be worse than the memory usage of just letting it grow to its maximum size, but we don't know until we measure, and it's probably not worth the trouble unless we start seeing it as a problem.